node/src/reactor/participating.rs

```diff
@@ -575,22 +569,73 @@ impl reactor::Reactor for Reactor {
             root,
             config,
             chainspec_loader,
-            mut storage,
+            storage,
             mut contract_runtime,
-            maybe_latest_block_header,
+            joining_outcome,
             event_stream_server,
             small_network_identity,
             node_startup_instant,
         } = config;
 
+        let effect_builder = EffectBuilder::new(event_queue);
+        let mut effects = Effects::new();
+        let latest_block_header = match joining_outcome {
+            JoiningOutcome::ShouldExitForUpgrade => {
+                error!("invalid joining outcome to transition to participating reactor");
+                return Err(Error::InvalidJoiningOutcome);
+            }
+            JoiningOutcome::Synced {
+                latest_block_header,
+            } => latest_block_header,
+            JoiningOutcome::RanUpgradeOrGenesis {
+                block_and_execution_effects:
+                    BlockAndExecutionEffects {
+                        block,
+                        execution_results,
+                        maybe_step_effect_and_upcoming_era_validators,
+                    },
+            } => {
+                // The outcome of joining in this case caused a new switch block to be created, so
+                // we need to emit the effects which would have been created by that execution, but
+                // add them to the participating reactor's event queues so they don't get dropped as
+                // the joining reactor shuts down.
+                effects.extend(
+                    effect_builder
+                        .announce_new_linear_chain_block(block.clone(), execution_results)
```

Comment:
We should check (maybe as a separate ticket) whether the era supervisor will actually create a finality signature for an immediate switch block.

Reply:
Sure - let's do that in a separate ticket